File: mobile-rn/src/components/ProductCard.js
import { Image, Linking, Pressable, StyleSheet, Text, View } from "react-native";
import { useAudioPlayer, useAudioPlayerStatus } from "expo-audio";
import { colors, radius } from "../theme";
import { api } from "../api";

function VoiceNoteButton({ audioUrl }) {
  const player = useAudioPlayer(audioUrl);
  const status = useAudioPlayerStatus(player);

  return (
    <Pressable style={styles.voiceBtn} onPress={() => (status.playing ? player.pause() : player.play())}>
      <Text style={styles.voiceBtnText}>{status.playing ? "⏸ Note vocale" : "▶ Écouter la note vocale"}</Text>
    </Pressable>
  );
}

function timeAgo(ts) {
  const s = Math.floor((Date.now() - ts) / 1000);
  if (s < 60) return "à l'instant";
  const m = Math.floor(s / 60);
  if (m < 60) return `il y a ${m} min`;
  const h = Math.floor(m / 60);
  if (h < 24) return `il y a ${h} h`;
  return `il y a ${Math.floor(h / 24)} j`;
}

function digitsOnly(str = "") {
  return str.replace(/[^\d]/g, "");
}

export default function ProductCard({ product: p, mine, onOpenChat, onOpenBoutique, onEdit, onDelete }) {
  const imageUrl = p.imageUrl.startsWith("http") ? p.imageUrl : `${api.base}${p.imageUrl}`;
  const audioUrl = p.audioUrl ? (p.audioUrl.startsWith("http") ? p.audioUrl : `${api.base}${p.audioUrl}`) : null;

  return (
    <View style={styles.card}>
      <View style={styles.photoWrap}>
        <Image source={{ uri: imageUrl }} style={styles.photo} />
        <View style={styles.vendorTag}>
          <Text style={styles.vendorTagText} numberOfLines={1}>
            {p.vendorId}
          </Text>
        </View>
      </View>

      <View style={styles.stub}>
        <Text style={styles.pname}>{p.name}</Text>
        <Text style={styles.pdesc}>{p.description}</Text>
        {!!audioUrl && <VoiceNoteButton audioUrl={audioUrl} />}

        {!mine && (
          <Pressable onPress={() => onOpenBoutique(p.sellerPhone, p.storeName || p.vendorId)}>
            <Text style={styles.vendorLine} numberOfLines={1}>
              🏪 {p.storeName || p.vendorId}
              {p.storeName ? ` · ${p.ownerLabel || p.vendorId}` : ""}
            </Text>
          </Pressable>
        )}

        <View style={styles.actions}>
          {mine ? (
            <>
              <Pressable style={styles.iconBtn} onPress={() => onEdit(p)}>
                <Text style={styles.iconBtnText}>✏️ Modifier</Text>
              </Pressable>
              <Pressable style={[styles.iconBtn, styles.deleteBtn]} onPress={() => onDelete(p.id)}>
                <Text style={[styles.iconBtnText, { color: colors.error }]}>🗑️ Supprimer</Text>
              </Pressable>
            </>
          ) : (
            <>
              <Pressable style={styles.iconBtn} onPress={() => onOpenChat(p)}>
                <Text style={styles.iconBtnText}>💬 Discuter</Text>
              </Pressable>
              {!!p.contact && (
                <Pressable
                  style={[styles.iconBtn, styles.whatsappBtn]}
                  onPress={() =>
                    Linking.openURL(
                      `https://wa.me/${digitsOnly(p.contact)}?text=${encodeURIComponent(
                        "Bonjour ! Je suis intéressé(e) par " + p.name + " sur Snapy."
                      )}`
                    )
                  }
                >
                  <Text style={[styles.iconBtnText, { color: colors.teal }]}>🟢 WhatsApp</Text>
                </Pressable>
              )}
            </>
          )}
        </View>
        <Text style={styles.ptime}>{timeAgo(p.createdAt)}</Text>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  card: {
    backgroundColor: colors.panel,
    borderRadius: radius,
    overflow: "hidden",
    borderWidth: 1,
    borderColor: colors.hairline,
    marginBottom: 16
  },
  photoWrap: { aspectRatio: 1, backgroundColor: "#0c0c12" },
  photo: { width: "100%", height: "100%" },
  vendorTag: {
    position: "absolute",
    top: 8,
    left: 8,
    backgroundColor: "rgba(16,15,22,0.75)",
    borderRadius: 20,
    paddingHorizontal: 8,
    paddingVertical: 3,
    maxWidth: "80%"
  },
  vendorTagText: { color: colors.teal, fontSize: 10 },
  stub: { padding: 14, borderTopWidth: 1, borderTopColor: colors.hairline, borderStyle: "dashed" },
  pname: { fontWeight: "700", fontSize: 15, color: colors.paper, marginBottom: 6 },
  pdesc: { fontSize: 12, lineHeight: 18, color: colors.textDim, marginBottom: 10 },
  vendorLine: { fontSize: 10.5, color: colors.teal, marginBottom: 10 },
  voiceBtn: {
    alignSelf: "flex-start",
    borderWidth: 1,
    borderColor: colors.hairline,
    borderRadius: 20,
    paddingHorizontal: 10,
    paddingVertical: 5,
    marginBottom: 10
  },
  voiceBtnText: { fontSize: 10.5, color: colors.teal },
  actions: { flexDirection: "row", flexWrap: "wrap", gap: 6, marginBottom: 8 },
  iconBtn: {
    borderWidth: 1,
    borderColor: colors.hairline,
    backgroundColor: colors.panelRaised,
    borderRadius: 20,
    paddingHorizontal: 10,
    paddingVertical: 6
  },
  iconBtnText: { fontSize: 10.5, color: colors.text },
  whatsappBtn: { borderColor: "rgba(95,195,176,0.35)" },
  deleteBtn: { borderColor: "rgba(232,117,106,0.35)" },
  ptime: { fontSize: 10, color: colors.hairline, marginTop: 6 }
});

File: mobile-flutter/lib/widgets/product_card.dart
import 'package:audioplayers/audioplayers.dart' as ap;
import 'package:flutter/material.dart';
import 'package:url_launcher/url_launcher.dart';

import '../api.dart';
import '../theme.dart';
import '../time_ago.dart';

class ProductCard extends StatelessWidget {
  final Product product;
  final bool mine;
  final void Function(Product) onOpenChat;
  final void Function(String phone, String label) onOpenBoutique;
  final void Function(Product) onOpenDetail;
  final void Function(Product) onEdit;
  final void Function(String id) onDelete;

  const ProductCard({
    super.key,
    required this.product,
    required this.mine,
    required this.onOpenChat,
    required this.onOpenBoutique,
    required this.onOpenDetail,
    required this.onEdit,
    required this.onDelete,
  });

  @override
  Widget build(BuildContext context) {
    final p = product;
    final imageUrl = p.absoluteImageUrl(Api.base);
    final audioUrl = p.absoluteAudioUrl(Api.base);

    return Container(
      margin: const EdgeInsets.only(bottom: 16),
      decoration: BoxDecoration(
        color: AppColors.panel,
        borderRadius: BorderRadius.circular(kRadius),
        border: Border.all(color: AppColors.hairline),
      ),
      clipBehavior: Clip.antiAlias,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          InkWell(
            onTap: () => onOpenDetail(p),
            child: Stack(
              children: [
                AspectRatio(
                  aspectRatio: 1,
                  child: Container(
                    color: const Color(0xFF0C0C12),
                    child: Image.network(imageUrl, fit: BoxFit.cover),
                  ),
                ),
                Positioned(
                  top: 8,
                  left: 8,
                  child: Container(
                    constraints: BoxConstraints(maxWidth: MediaQuery.of(context).size.width * 0.6),
                    padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 3),
                    decoration: BoxDecoration(color: const Color(0xBF100F16), borderRadius: BorderRadius.circular(20)),
                    child: Text(p.vendorId, maxLines: 1, overflow: TextOverflow.ellipsis, style: const TextStyle(color: AppColors.teal, fontSize: 10)),
                  ),
                ),
              ],
            ),
          ),
          Container(
            padding: const EdgeInsets.all(14),
            decoration: const BoxDecoration(border: Border(top: BorderSide(color: AppColors.hairline))),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                GestureDetector(
                  onTap: () => onOpenDetail(p),
                  child: Text(p.name, style: const TextStyle(fontWeight: FontWeight.w700, fontSize: 15, color: AppColors.paper)),
                ),
                const SizedBox(height: 6),
                Text(p.description, style: const TextStyle(fontSize: 12, height: 1.5, color: AppColors.textDim)),
                if (audioUrl != null) ...[
                  const SizedBox(height: 8),
                  _VoiceNoteButton(audioUrl: audioUrl),
                ],
                const SizedBox(height: 8),
                GestureDetector(
                  onTap: () => onOpenDetail(p),
                  child: const Text('Plus →', style: TextStyle(fontSize: 10.5, color: AppColors.amber)),
                ),
                const SizedBox(height: 10),
                if (!mine)
                  GestureDetector(
                    onTap: () => onOpenBoutique(p.sellerPhone, p.storeName.isNotEmpty ? p.storeName : p.vendorId),
                    child: Padding(
                      padding: const EdgeInsets.only(bottom: 10),
                      child: Text(
                        '🏪 ${p.storeName.isNotEmpty ? p.storeName : p.vendorId}${p.storeName.isNotEmpty ? ' · ${p.ownerLabel.isNotEmpty ? p.ownerLabel : p.vendorId}' : ''}',
                        maxLines: 1,
                        overflow: TextOverflow.ellipsis,
                        style: const TextStyle(fontSize: 10.5, color: AppColors.teal),
                      ),
                    ),
                  ),
                Wrap(
                  spacing: 6,
                  runSpacing: 6,
                  children: mine
                      ? [
                          _ActionChip(label: '✏️ Modifier', onTap: () => onEdit(p)),
                          _ActionChip(label: '🗑️ Supprimer', color: AppColors.error, borderColor: const Color(0x59E8756A), onTap: () => onDelete(p.id)),
                        ]
                      : [
                          _ActionChip(label: '💬 Discuter', onTap: () => onOpenChat(p)),
                          if (p.contact.isNotEmpty)
                            _ActionChip(
                              label: '🟢 WhatsApp',
                              color: AppColors.teal,
                              borderColor: const Color(0x595FC3B0),
                              onTap: () => launchUrl(
                                Uri.parse(
                                  'https://wa.me/${digitsOnly(p.contact)}?text=${Uri.encodeComponent("Bonjour ! Je suis intéressé(e) par ${p.name} sur Snapy.")}',
                                ),
                                mode: LaunchMode.externalApplication,
                              ),
                            ),
                        ],
                ),
                const SizedBox(height: 6),
                Text(timeAgo(p.createdAt), style: const TextStyle(fontSize: 10, color: AppColors.hairline)),
              ],
            ),
          ),
        ],
      ),
    );
  }
}

class _VoiceNoteButton extends StatefulWidget {
  final String audioUrl;

  const _VoiceNoteButton({required this.audioUrl});

  @override
  State<_VoiceNoteButton> createState() => _VoiceNoteButtonState();
}

class _VoiceNoteButtonState extends State<_VoiceNoteButton> {
  final _player = ap.AudioPlayer();
  bool _playing = false;

  @override
  void initState() {
    super.initState();
    _player.onPlayerStateChanged.listen((state) {
      if (mounted) setState(() => _playing = state == ap.PlayerState.playing);
    });
  }

  @override
  void dispose() {
    _player.dispose();
    super.dispose();
  }

  Future<void> _toggle() async {
    if (_playing) {
      await _player.pause();
    } else {
      await _player.play(ap.UrlSource(widget.audioUrl));
    }
  }

  @override
  Widget build(BuildContext context) {
    return InkWell(
      onTap: _toggle,
      borderRadius: BorderRadius.circular(20),
      child: Container(
        padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 6),
        decoration: BoxDecoration(
          color: AppColors.panelRaised,
          borderRadius: BorderRadius.circular(20),
          border: Border.all(color: AppColors.hairline),
        ),
        child: Text(_playing ? '⏸ Note vocale' : '▶ Écouter la note vocale', style: const TextStyle(fontSize: 10.5, color: AppColors.teal)),
      ),
    );
  }
}

class _ActionChip extends StatelessWidget {
  final String label;
  final Color? color;
  final Color? borderColor;
  final VoidCallback onTap;

  const _ActionChip({required this.label, required this.onTap, this.color, this.borderColor});

  @override
  Widget build(BuildContext context) {
    return InkWell(
      onTap: onTap,
      borderRadius: BorderRadius.circular(20),
      child: Container(
        padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 6),
        decoration: BoxDecoration(
          color: AppColors.panelRaised,
          borderRadius: BorderRadius.circular(20),
          border: Border.all(color: borderColor ?? AppColors.hairline),
        ),
        child: Text(label, style: TextStyle(fontSize: 10.5, color: color ?? AppColors.text)),
      ),
    );
  }
}
